engine/src/main/java/org/hibernate/validator/internal/engine/tracking/PredefinedScopeProcessedBeansTrackingStrategy.java

```diff
@@ -0,0 +1,368 @@
+/*
+ * SPDX-License-Identifier: Apache-2.0
+ * Copyright Red Hat Inc. and Hibernate Authors
+ */
+package org.hibernate.validator.internal.engine.tracking;
+
+import java.lang.reflect.ParameterizedType;
+import java.lang.reflect.Type;
+import java.lang.reflect.TypeVariable;
+import java.lang.reflect.WildcardType;
+import java.util.HashMap;
+import java.util.HashSet;
+import java.util.Map;
+import java.util.Set;
+
+import org.hibernate.validator.internal.engine.valueextraction.ValueExtractorDescriptor;
+import org.hibernate.validator.internal.metadata.aggregated.BeanMetaData;
+import org.hibernate.validator.internal.metadata.aggregated.CascadingMetaData;
+import org.hibernate.validator.internal.metadata.aggregated.ContainerCascadingMetaData;
+import org.hibernate.validator.internal.metadata.aggregated.PotentiallyContainerCascadingMetaData;
+import org.hibernate.validator.internal.metadata.aggregated.ReturnValueMetaData;
+import org.hibernate.validator.internal.metadata.aggregated.ValidatableParametersMetaData;
+import org.hibernate.validator.internal.metadata.facets.Cascadable;
+import org.hibernate.validator.internal.properties.Signature;
+import org.hibernate.validator.internal.util.CollectionHelper;
+
+public class PredefinedScopeProcessedBeansTrackingStrategy implements ProcessedBeansTrackingStrategy {
+
+	private final Map<Class<?>, Boolean> trackingEnabledForBeans;
+
+	// TODO: signature is just name and parameters so that can clash between different beans.
+	//  with that.. do we even need to track it per signature or since
+	//  we already built the `trackingEnabledForBeans` we can just "inspect" the cascadable as we go
+	//  and check against this `trackingEnabledForBeans` to see if tracking is required ?
+	private final Map<Signature, Boolean> trackingEnabledForReturnValues;
+	private final Map<Signature, Boolean> trackingEnabledForParameters;
```

Comment:
I'm not sure we need to prebuild these maps, but I wanted to check with you first before dropping them...
As per that TODO, it seems that we would only check an executable once when building the metadata and I think I got an impl that would rely only on the `trackingEnabledForBeans` ...  
and since the signatures could clash ... I guess we can drop these maps, right?